pennylane/devices/qubit/simulate.py

```diff
@@ -316,6 +347,334 @@ def simulate_partial(k):
+
+
+def prune_mcm_samples(op, branch, mcm_active, mcm_samples):
+    """Removes samples from mid-measurement sample dictionary given a MidMeasureMP and branch.
+
+    Post-selection on a given mid-circuit measurement leads to ignoring certain branches
+    of the tree and samples. The corresponding samples in all other mid-circuit measurement
+    must be deleted accordingly. We need to find which samples are
+    corresponding to the current branch by looking at all parent nodes.
+    """
+    mask = mcm_samples[op] == branch
+    for k, v in mcm_active.items():
+        if k == op:
+            break
+        mask = np.logical_and(mask, mcm_samples[k] == v)
+    for k in mcm_samples.keys():
+        mcm_samples[k] = mcm_samples[k][np.logical_not(mask)]
+
+
+def update_mcm_samples(op, samples, mcm_active, mcm_samples):
+    """Updates the mid-measurement sample dictionary given a MidMeasureMP and samples.
+
+    If the ``mcm_active`` dictionary is empty, we are at the root and ``mcm_samples`
+    is simply updated with ``samples``.
+
+    If the ``mcm_active`` dictionary is not empty, we need to find which samples are
+    corresponding to the current branch by looking at all parent nodes. ``mcm_samples`
+    is then updated with samples at indices corresponding to parent nodes.
+    """
+    if mcm_active:
+        shape = next(iter(mcm_samples.values())).shape
+        mask = np.ones(shape, dtype=bool)
+        for k, v in mcm_active.items():
+            if k == op:
+                break
+            mask = np.logical_and(mask, mcm_samples[k] == v)
+        if op not in mcm_samples:
+            mcm_samples[op] = np.empty(shape, dtype=samples.dtype)
+        mcm_samples[op][mask] = samples
+    else:
+        mcm_samples[op] = samples
+
+
+def circuit_up_to_first_mcm(circuit):
+    """Returns two circuits; one that runs up-to the next mid-circuit measurement and one that runs beyond it."""
+    if not has_mid_circuit_measurements(circuit):
+        return circuit, None, None
+
+    # find next MidMeasureMP
+    def find_next_mcm(circuit):
+        for i, op in enumerate(circuit.operations):
+            if isinstance(op, MidMeasureMP):
+                return i, op
+        return len(circuit.operations) + 1, None
+
+    i, op = find_next_mcm(circuit)
+    # run circuit until next MidMeasureMP and sample
+    circuit_base = qml.tape.QuantumScript(
+        circuit.operations[0:i],
+        [qml.sample(wires=op.wires) if op.obs is None else qml.sample(op=op.obs)],
+        shots=circuit.shots,
+        trainable_params=circuit.trainable_params,
+    )
+    # circuit beyond next MidMeasureMP with VarianceMP <==> SampleMP
+    new_measurements = []
+    for m in circuit.measurements:
+        if not m.mv:
+            if isinstance(m, VarianceMP):
+                new_measurements.append(SampleMP(obs=m.obs))
+            else:
+                new_measurements.append(m)
```

Comment:
It is replaced for terminal `var` measurements of MCMs. Think of the extreme case where we end up with a single shot per explored branch. The variance of a single-element array is always `0`, so we cannot combine local variances into a global variance as we can for other measurement processes. We thus request `sample(mcm)` and compute its variance once we have all samples.

Reply:
But the if statements means the variance is only swapped when it _doesnt_ act on an mcm no?

Reply:
Sorry, you are correct. I think the issue is we need the global `expval` (or `sample`) to compute the variance, but we do not have it until the end. That is true regardless of the observable, and hence we replace the terminal measurement by `sample` and compute the variance once we have the global value.

Reply:
Are you saying variances cannot be computed at the end of a specific branch, while all other MPs can immediately be computed from the obtained samples on a specific branch?

Reply:
Yes.

Reply:
Alright so a summary of how this works:
- generally MPs are computed partially on each branch, and then combined at the end (for example a weighted sum for expval)
- for variance there is no way to recombine "partial variances" from each branch so it needs to be handled once samples are available from all branches

I have to say I find the partial MP approach a bit more confusing, in particular since Var is excluded, and it might be slightly less efficient than computing MPs in one go from all samples, but since you say you had issues delaying other MPs we can leave it as it.

It would be great to add a comment though about why Var is excluded (and how its computation differs from that of the other MPs).